File: services/payment-service/internal/model/payment.go
package model

import (
	"time"
)

type PaymentStatus string

const (
	StatusPending   PaymentStatus = "pending"
	StatusCompleted PaymentStatus = "completed"
	StatusFailed    PaymentStatus = "failed"
	StatusCancelled PaymentStatus = "cancelled"
)

type Payment struct {
	ID            string        `json:"id"`
	AppointmentID string        `json:"appointment_id"`
	PatientID     string        `json:"patient_id"`
	Amount        float64       `json:"amount"`
	Currency      string        `json:"currency"`
	Status        PaymentStatus `json:"status"`
	Provider      string        `json:"provider"` // e.g., "payhere"
	ProviderID    string        `json:"provider_id"`
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`
}

type CreatePaymentRequest struct {
	AppointmentID string  `json:"appointment_id" binding:"required,uuid"`
	PatientID     string  `json:"patient_id" binding:"required,uuid"`
	Amount        float64 `json:"amount" binding:"required,gt=0"`
	Currency      string  `json:"currency" binding:"required"`
}

type PaymentResponse struct {
	PaymentID string `json:"payment_id"`
	Status    string `json:"status"`
	Provider  string `json:"provider,omitempty"`
}

type CheckoutRequest struct {
	PaymentID     string          `json:"payment_id" binding:"omitempty,uuid"`
	AppointmentID string          `json:"appointment_id" binding:"omitempty,uuid"`
	Items         string          `json:"items"`
	Customer      PayHereCustomer `json:"customer"`
}

type PayHereCustomer struct {
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Email     string `json:"email"`
	Phone     string `json:"phone"`
	Address   string `json:"address"`
	City      string `json:"city"`
	Country   string `json:"country"`
}

type CheckoutResponse struct {
	PaymentID   string            `json:"payment_id"`
	Provider    string            `json:"provider"`
	CheckoutURL string            `json:"checkout_url"`
	Fields      map[string]string `json:"fields"`
}

// PayHereNotification is the POST payload sent to your notify_url (webhook).
// PayHere sends it as x-www-form-urlencoded.
type PayHereNotification struct {
	MerchantID      string `form:"merchant_id" json:"merchant_id"`
	OrderID         string `form:"order_id" json:"order_id"`
	PaymentID       string `form:"payment_id" json:"payment_id"`
	PayHereAmount   string `form:"payhere_amount" json:"payhere_amount"`
	PayHereCurrency string `form:"payhere_currency" json:"payhere_currency"`
	StatusCode      int    `form:"status_code" json:"status_code"`
	MD5Sig          string `form:"md5sig" json:"md5sig"`
	Custom1         string `form:"custom_1" json:"custom_1"`
	Custom2         string `form:"custom_2" json:"custom_2"`
	StatusMessage   string `form:"status_message" json:"status_message"`
}

File: frontend/src/api/payment.ts
import apiClient from './client';

export interface CheckoutRequest {
    payment_id?: string;
    appointment_id: string;
    patient_id: string;
    amount: number;
    currency: string;
    items: string;
    customer: {
        first_name: string;
        last_name: string;
        email: string;
        phone: string;
        address: string;
        city: string;
        country: string;
    };
}

export interface CheckoutResponse {
    payment_id: string;
    provider: string;
    checkout_url: string;
    fields: Record<string, string>;
}

export interface Payment {
    id: string;
    appointment_id: string;
    patient_id: string;
    amount: number;
    currency: string;
    status: 'pending' | 'completed' | 'failed' | 'refunded' | 'partially_refunded';
    provider: string;
    provider_id: string;
    created_at: string;
    updated_at: string;
}

export const paymentApi = {
    createPayment: async (data: {
        appointment_id: string;
        patient_id: string;
        amount: number;
        currency: string;
    }) => {
        const response = await apiClient.post<{ payment_id: string; status: string; provider?: string }>('v1/payments/', data);
        return response.data;
    },

    checkout: async (data: CheckoutRequest): Promise<CheckoutResponse> => {
        const response = await apiClient.post<CheckoutResponse>('v1/payments/checkout', data);
        return response.data;
    },

    completePayment: async (paymentId: string) => {
        const response = await apiClient.post(`v1/payments/${paymentId}/complete`);
        return response.data;
    },
    
    getPaymentStatus: async (paymentId: string) => {
        const response = await apiClient.get<Payment>(`v1/payments/${paymentId}`);
        return response.data;
    },

    listPayments: async (patientId: string) => {
        const response = await apiClient.get<Payment[]>(`v1/payments/patient/${patientId}`);
        return response.data;
    }
};
